code.rb

```diff
@@ -0,0 +1,16 @@
+class Code
+  DEST_CODE = {
+    "null" => "000",
+    "M" => "001",
+    "D" => "010",
+    "MD" => "011",
+    "A" => "100",
+    "AM" => "101",
+    "AD" => "110",
+    "AMD" => "111",
+  }
+
+  def dest(mnemonic)
+    DEST_CODE[mnemonic]
+  end
```

Comment:
Nice!

You also suggested an alternative strategy that generated the three bits based on what letters were in the mnemonic. Here’s what that looks like, if you’re interested:

```ruby
def dest(mnemonic)
  'ADM'
    .chars
    .map { |letter| mnemonic.include?(letter) }
    .map { |on| on ? '1' : '0' }
    .join
end
```

Do you like this better or worse than the hash? What are the pros and cons?

Reply:
* Pros: great use of Ruby's method chaining, less code compared to having an entire hash table, better scalability/reusability [1]
* Cons: not too obvious at first glance what the code is doing... that's it.

I don't have a strong preference, and I may even prefer the hash table more because it's obvious what we're trying to do. What do you think? 

[1] Can reuse the chaining logic because the only variance is the mnemonic format.

Reply:
We discussed this verbally, but for the record:

That’s a great analysis! I agree with you that the hash table is probably more appropriate in this situation because it’s already so short, simple, fast and easy to understand. The “clever” code is more concise and perhaps cooler, but it’s definitely harder to see what’s going on at a glance.

A hypothetical advantage of the `'ADM'.chars.map` version is that it’d be easier to extend in the future if requirements changed. If an imaginary Hack v2 processor was announced with six registers instead of three, it would be easy to just add the new register characters onto the end of the `'ADM'` string to make the method begin returning six bits instead of three. Whereas the hash table would have to grow to 8 times the size to support the same requirement, and the chances of it containing an error would increase. So at some point in the future the hash table might stop being the more convenient option, but spotting exactly when that point arrives is more of an art than a science.

Another more abstract advantage of the chained version is that, because it’s a _general_ solution to the problem, it’s easier to convince ourselves that it will work in all cases we care about. We can probably pick a few arbitrary test cases (e.g. try calling it with `'null'`, `'AD'` and `'AMD'`) and if they all pass then we can be pretty confident that all the untested cases will work too, because they’re all produced by the same logic — if there was a problem with the code, we’d probably have discovered it with a few representative tests. Whereas the big hash is, well, just a big hash — just because we’ve checked that a few of the values in it are correct, that doesn’t necessarily mean that the other ones are. The only way to be _really_ sure is to test all of them, but that feels… kind of pointless? So there’s at least some advantage in writing code that actively “solves the problem” rather than just looking up the answers in a big table, because it’s harder to be sure that every entry in the big table is necessarily correct.